components/sup/src/manager/file_watcher.rs

```diff
@@ -26,3439 +31,1461 @@ use std::{collections::{hash_map::Entry,
-                                               self.debug_info,)
-                                    });
             }
-            event_count
         }
+    }
 
-        fn get_event_count_on_rm_rf(&self, top_path: &Path) -> u32 {
-            let mut queue = VecDeque::new();
-            queue.push_back(top_path.to_path_buf());
-            let mut event_count = 0;
-            while let Some(path) = queue.pop_front() {
-                if !self.parent_is_watched(&path) {
-                    continue;
-                }
-                // Two events for each deletion in the watched path -
-                // remove notice and remove.
-                event_count += 2;
-                let metadata = match path.symlink_metadata() {
-                    Ok(m) => m,
-                    Err(err) => {
-                        panic!("Failed to stat {}: {}, debug info:\n{}",
-                               path.display(),
-                               err,
-                               self.debug_info,)
+    fn handle_folder_disappearance(&mut self, notify_event: &Event) {
+        trace!("FileWatcher::handle_folder_disappearance\n-=> {:?}\n-=> {:?}\n",
+               notify_event,
+               self.watched_file);
+        if self.watched_file.state != WatchedFileState::NonExistent {
+            for path_buf in notify_event.paths.iter() {
+                let pb = match path_buf.canonicalize() {
+                    Ok(v) => v,
+                    Err(e) => {
+                        debug!("failed canonicalization: {:?}", e);
+                        path_buf.to_path_buf()
                     }
                 };
-                if !metadata.is_dir() {
-                    continue;
+                if self.watched_file.path == pb || self.watched_file.directory == pb {
+                    match self.manage_watches() {
+                        Ok(_) => {
+                            break;
+                        }
+                        Err(e) => {
+                            debug!("While managing watches: {:?}", e);
+                            self.log_ignored_path(notify_event, path_buf)
+                        }
+                    }
+                } else {
+                    debug!("fell through");
+                    self.log_ignored_path(notify_event, path_buf);
                 }
-                queue.extend(self.get_dir_contents(&path));
             }
-            event_count
         }
+    }
+
+    // Logging that a particular event was ignored.  Ignoring an event may be
+    // the correct thing to do so the purpose of logging that a particular event
+    // was ignored event is to highlight the fact that an event was ignored in
+    // case the decision made today to ignore that event needs to be changed
+    // due to design changes, lessons learned (bugs) or notify crate updates.
+    fn log_ignored_event(&mut self, notify_event: &Event) { trace!("IGNORING {:?}", notify_event) }
+
```

Comment:
How frequent will this be logged? I realize it's trace but I'm just curious if this will become a firehose of unneeded logging.

Reply:
We talked about this at one point but capturing the conversation I don't expect this to be a frequently logged item. I do think if anyone is trying to figure out why a watch isn't alerting that having them to turn on trace and seeing this would quickly help pinpoint why.